### PR TITLE
feat/rudderstack payload changes

### PR DESCRIPTION
**Pull request by amandal97**
## Description
The changes include `rudderstack` api payload fixes.
1. [JIRA](https://razorpay.atlassian.net/browse/OBSV-1177?atlOrigin=eyJpIjoiMjEyMGYwNzRmZDQ4NGQzY2E4MmQ1OThkNWRjZTc0ODQiLCJwIjoiaiJ9) | `userInfo` property should have the value returned from `/user-info` api call. The solution was to store the data in `local-storage` so that the data is not flushed out everytime the page is getting refreshed. Also, the api call is being cancelled if the data is already present in `local-storage`. 
2. [JIRA](https://razorpay.atlassian.net/browse/OBSV-1178?atlOrigin=eyJpIjoiZmRhNGI4OWIxM2YwNDY2YzkxODlhYjJhZjFmMjExNzIiLCJwIjoiaiJ9) | `diffInDays` should be provided when custom time range is selected. It is the no. of days between current time and the start time provided in the custom time range.

### Testing
Tested in local

### Checklist:
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: projects/common/src/telemetry/user-telemetry-impl.service.ts
Comment (cseas) [inline projects/common/src/telemetry/user-telemetry-impl.service.ts]:
```suggestion
  private getDifferenceInDays(timeParamValue: string | undefined): number | undefined {
```

File: projects/common/src/telemetry/user-telemetry-impl.service.ts
Comment (cseas) [inline projects/common/src/telemetry/user-telemetry-impl.service.ts]: When is it undefined?

File: projects/common/src/telemetry/user-telemetry-impl.service.ts
Comment (cseas) [inline projects/common/src/telemetry/user-telemetry-impl.service.ts]:
```suggestion
      const differenceInMilliseconds = endDate.getTime() - startTime;
```

File: projects/common/src/telemetry/user-telemetry-impl.service.ts
Comment (cseas) [inline projects/common/src/telemetry/user-telemetry-impl.service.ts]:
```suggestion
      return parseInt(diffInDays);
```

File: projects/common/src/telemetry/user-telemetry-impl.service.ts
Comment (amandal97) [reply projects/common/src/telemetry/user-telemetry-impl.service.ts]: This is for `null` check. I cannot think of any scenario where this will come as `null`, but the type definition of `timeParamValue` is `string | undefined | null`. 

Have removed the `null` check and only kept `undefined` check as it caused linting issues:
Screenshot 2022-12-13 at 4 57 43 PM

File: projects/common/src/telemetry/user-telemetry-impl.service.ts
Comment (amandal97) [reply projects/common/src/telemetry/user-telemetry-impl.service.ts]: Have made the change. Please check.

File: projects/common/src/telemetry/user-telemetry-impl.service.ts
Comment (cseas) [reply projects/common/src/telemetry/user-telemetry-impl.service.ts]: @jaywalker21 We should probably remove this rule. What do you think?
`null` is intentional. `undefined` usually means a bug in most codebases.

File: projects/common/src/telemetry/user-telemetry-impl.service.ts
Comment (amandal97) [reply projects/common/src/telemetry/user-telemetry-impl.service.ts]: The above will not work as parseInt expects `string` as its argument.
Therefore used `toFixed()` to remove the decimal numbers which converts it to string and henceforth used Number() to convert it back to `number`.

Also, can use `parseInt(diffInDays.toFixed(0), 10)`

File: projects/common/src/telemetry/user-telemetry-impl.service.ts
Comment (amandal97) [reply projects/common/src/telemetry/user-telemetry-impl.service.ts]: Made the changes. Please check.

File: projects/common/src/telemetry/user-telemetry-impl.service.ts
Comment (cseas) [reply projects/common/src/telemetry/user-telemetry-impl.service.ts]: Working for both strings and numbers in the browser console. Is the code throwing error?

Screenshot 2022-12-13 at 6 10 37 PM

Comment (cseas) [reply projects/common/src/telemetry/user-telemetry-impl.service.ts]: Just checked, typescript is expecting a string. Nevermind, ignore the comment.

**Review by cseas — APPROVED** [review]
LGTM

File: projects/common/src/telemetry/user-telemetry-impl.service.ts
Comment (cseas) [inline projects/common/src/telemetry/user-telemetry-impl.service.ts]: Not a blocker, but try adding a test for this function before merging to increase coverage a bit.

File: projects/common/src/telemetry/user-telemetry-impl.service.ts
Comment (jaywalker21) [reply projects/common/src/telemetry/user-telemetry-impl.service.ts]: I don't think so tbh, Consider methods built in javascript like [find](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find), they intentionally return `undefined` when no matching element is found, they could have returned `null`.

To me, `null` makes more sense when you absolutely want to reset some value. Also [look](https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines#null-and-undefined) what Typescript core team has to say on it.

So, based on the above links, I strongly feel that the rule is ideal here.

Comment (jaywalker21) [reply projects/common/src/telemetry/user-telemetry-impl.service.ts]: @amandal97 It can be `null` or `undefined` too. Say you launch, `localhost:4200/home`.
At that time the value would not be there. That is the reason, it's type is a union :)

File: projects/common/src/telemetry/user-telemetry-impl.service.ts
Comment (jaywalker21) [inline projects/common/src/telemetry/user-telemetry-impl.service.ts]: I think we can leverage the `getDate` method and get the difference in days.

```suggestion
      const timeRangeArray = /(\d+)-(\d+)/.exec(timeRangeValue);
      const todayDate = new Date();
      const searchStartDate = new Date(+timeRangeArray[1]);
      return todayDate.getDate() - searchStartDate.getDate();
```

**Review by jaywalker21 — COMMENTED** [review]
1-2 comments, rest all looks good.

File: projects/common/src/telemetry/user-telemetry-impl.service.ts
Comment (amandal97) [reply projects/common/src/telemetry/user-telemetry-impl.service.ts]: Understood @jaywalker21

File: projects/common/src/telemetry/user-telemetry-impl.service.ts
Comment (amandal97) [reply projects/common/src/telemetry/user-telemetry-impl.service.ts]: Done.

**Review by jaywalker21 — APPROVED** [review]
Looks good to me

**Pull request merged**
